### PR TITLE
Make logsize settable for each action

### PR DESCRIPTION
**Pull request by markusthoemmes**
Adds the logsize as a limit that is settable per action and thus included in the quota for each action.

@mdeuser please review the go-cli parts

Signed-off-by: Christian Bickel <cbickel@de.ibm.com>

### CODE REVIEW

**Comment by markusthoemmes** [comment]
PPG 897

File: tools/go-cli/go-whisk/whisk/shared.go [hunk 1]
```diff
@@ -38,4 +38,5 @@ type Parameters *json.RawMessage
 type Limits struct {
     Timeout int `json:"timeout,omitempty"`
     Memory  int `json:"memory,omitempty"`
+    Logsize int `json:"logs"`
```
Comment (mdeuser) [inline tools/go-cli/go-whisk/whisk/shared.go]: IMHO, the JSON label "logsize" would better fit the purpose.
Add "omitempty" so that a `wsk action update` won't accidentally reset this value to the 0

Comment (markusthoemmes) [inline tools/go-cli/go-whisk/whisk/shared.go]: `omitempty` is not feasible though, it will turn a `0` (perfectly valid) to an empty object. Have a solution for that?

**Comment by mdeuser** [comment]
Are there any doc updates providing more details about this feature?
- What happens when the log limit is reached (i.e. log file wraps, no logging, log file is reset, etc)?
- Default log size when none is set?

**Comment by rabbah** [comment]
This is a new limit we need do document - @csantana FYI.

File: tools/go-cli/go-whisk/whisk/shared.go (hunk 1 above)
Comment (mdeuser) [inline tools/go-cli/go-whisk/whisk/shared.go]: Ah.. The default values have been changed from 0 to -1.  What will the server do with -1 values on both `wsk action create` and `wsk action update`?

Comment (markusthoemmes) [inline tools/go-cli/go-whisk/whisk/shared.go]: @ChrBi and me played a bit with it. They are not sent at all (as they weren't with `0` set). Note that the server will decline `0` for memory and timeout limits.

Comment (mdeuser) [inline tools/go-cli/go-whisk/whisk/shared.go]: The limits struct is instantiated when any of the memory, timeout or logsize arg is used.  Any arg not set will default to 0.  When the `omitempty` is removed, the default 0 value will be sent.  Can you try this test.
```
touch a.js
wsk action create tempaction a.js --logsize 10
wsk action get tempaction
wsk action update tempaction --memory 512  // leaving logsize alone
wsk action get tempaction    // Did the logsize setting change?
```

Comment (markusthoemmes) [inline tools/go-cli/go-whisk/whisk/shared.go]: Cheers for pointing that out. We haven't tried that. The unfortunate is: `0` is overloaded as the "empty" value for `int`. Investigating...

File: docs/reference.md [hunk 2]
```diff
@@ -420,6 +421,11 @@ OpenWhisk has a few system limits, including how much memory an action uses and
 * A user can change the limit when creating the action.
 * A container cannot have more memory allocated than the limit.
 
+### Per action logs (MB) (Default: 10MB)
+* The log limit M is in the range from [0MB..10MB] and is set per action in MB.
+* A user can change the limit when creating the action.
+* If a container writes more logs than the limit, the activation will contain the logs until that limit and truncated afterwards. The last logline will contain warning.
```
Comment (mdeuser) [inline docs/reference.md]: - Just confirming.  "truncated afterward" means that no additional log entries will be added to the log file ?
- What happens when the size is set to 0MB?

Comment (markusthoemmes) [inline docs/reference.md]: 1. Right, my wording might be off, suggestions?
2. Setting it to 0 will either result in an empty log (nothing written) or in just the warning of truncation

Comment (rabbah) [inline docs/reference.md]: While updating the limits, should also add the bits for
exec/annotations/parameters.

File: docs/reference.md [hunk 3]
```diff
@@ -406,6 +406,7 @@ OpenWhisk has a few system limits, including how much memory an action uses and
 | ----- | ----------- | ------------ | -----| ------- |
 | timeout | a container is not allowed to run longer than N milliseconds | per action |  milliseconds | 60000 |
 | memory | a container is not allowed to allocate more than N MB of memory | per action | MB | 256 |
+| logs | a container is not allowed to write mor than N MB to stdout | per action | MB | 10 |
```
Comment (rabbah) [inline docs/reference.md]: mor -> more

File: docs/reference.md [hunk 4]
```diff
@@ -420,6 +421,11 @@ OpenWhisk has a few system limits, including how much memory an action uses and
 * A user can change the limit when creating the action.
 * A container cannot have more memory allocated than the limit.
 
+### Per action logs (MB) (Default: 10MB)
+* The log limit M is in the range from [0MB..10MB] and is set per action in MB.
```
Comment (rabbah) [inline docs/reference.md]: using M here, N above. suggest this:
 +* The log limit N is in the range [0MB..10MB] and is set per action.

File: docs/reference.md [hunk 5]
```diff
@@ -420,6 +421,11 @@ OpenWhisk has a few system limits, including how much memory an action uses and
...
+* A user can change the limit when creating the action.
```
Comment (rabbah) [inline docs/reference.md]: or updating.

File: docs/reference.md (hunk 2 above)
Comment (rabbah) [inline docs/reference.md]: suggested alternative: Logs that exceed the set limit are truncated and a warning is added as the last output of the activation to indicate that the activation exceeded the set log limit.

File: tests/src/common/Wsk.scala [hunk 6]
```diff
@@ -226,6 +225,7 @@ class WskAction(override val usePythonCLI: Boolean = false)
         annotations: Map[String, JsValue] = Map(),
         timeout: Option[Duration] = None,
         memory: Option[Int] = None,
+        logsize: Option[Int] = None,
```
Comment (rabbah) [inline tests/src/common/Wsk.scala]: you could make this take Size instead of Int (since you're using `0 MB` in the test with `.toInt`.

File: tools/cli/wskaction.py [hunk 7]
```diff
@@ -95,7 +97,8 @@ def create(self, args, props, update):
             if args.param:
                 payload['parameters'] = getParams(args)
             # API will accept limits == {} as limits not specified on an update
-            if args.timeout or args.memory:
+            print args.logsize
```
Comment (rabbah) [inline tools/cli/wskaction.py]: remove.

File: tools/cli/wskaction.py [hunk 8]
```diff
@@ -95,7 +97,8 @@ def create(self, args, props, update):
...
+            if args.timeout or args.memory or not args.logsize is None:
```
Comment (rabbah) [inline tools/cli/wskaction.py]: should these all be `is None` vs falsy comparisons?

**Comment by rabbah** [comment]
cli tests for the corner cases (create action and get limits sufficient to make sure covered).

File: tools/go-cli/go-whisk-cli/commands/action.go [hunk 9]
```diff
@@ -506,13 +506,22 @@ func parseAction(cmd *cobra.Command, args []string) (*whisk.Action, bool, error)
...
     // Only include the memory and timeout limit if set
-    if flags.action.memory > 0 || flags.action.timeout > 0 {
+    if flags.action.memory > -1 || flags.action.timeout > -1 || flags.action.logsize > -1 {
         limits = new(whisk.Limits)
-        if flags.action.memory > 0 {
-            limits.Memory = flags.action.memory
+        if flags.action.memory > -1 {
+            limits.Memory = &flags.action.memory
+        } else {
+            limits.Memory = nil
```
Comment (mdeuser) [inline tools/go-cli/go-whisk-cli/commands/action.go]: Allocated pointers are initialized to nil, so there's actually no need to explicitly set these limits.* pointers to nil.

Comment (markusthoemmes) [inline tools/go-cli/go-whisk-cli/commands/action.go]: Makes sense, thanks!

**Comment by rabbah** [comment]
Ready modulo corner case tests for CLI which I volunteered to write.

File: docs/reference.md [hunk 10]
```diff
@@ -420,6 +421,11 @@ OpenWhisk has a few system limits, including how much memory an action uses and
 * A user can change the limit when creating the action.
 * A container cannot have more memory allocated than the limit.
 
+### Per action logs (MB) (Default: 10MB)
+* The log limit N is in the range [0MB..10MB] and is set per action.
+* A user can change the limit when creating or updating the action.
+* Logs that exceed the set limit are truncated and a warning is added as the last output of the activation to indicate that the activation exceeded the set log limit.
+
```
Comment (mdeuser) [inline docs/reference.md]: Add doc clarifying behavior of log when size limit is set to 0.

**Comment by rabbah** [comment]
Squashed rebased and merged as https://github.com/openwhisk/openwhisk/commit/f0dab306bd8e7bfc9fcafde0467af985940131c9.

**Comment by rabbah** [comment]
@mdeuser I thought it was clear what happens (ie the logs are truncated even with 0MB) - do you have suggested alternative language?

**Comment by mdeuser** [comment]
Since a 0MB log size limit is an unusual setting, I thought it might need explicit doc.  In other products, I've seen the use of 0 to mean "unlimited".
